### PR TITLE
Add quorum and BDFL notes to the CoC

### PR DESCRIPTION
**Pull request by iravid**

### CODE REVIEW

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -6,14 +6,31 @@ identity and expression, sexual orientation, disability, personal
...
+The ZIO project follows the [Scala Code of Conduct](https://www.scala-lang.org/conduct/), with
+adjustments with clarifications with regard to moderation that are detailed below.
+All participants, contributors and members are expected to follow the Scala Code of Conduct
+when discussing the project on the available communication channels. If you are being harassed, 
+please contact us immediately so that we can support you.
 
- ## Moderation
+## Moderation and Steering Committee
+ 
+The ZIO project's group of moderators is the Steering Committee team. The members currently are:
```
Comment (mijicd) [inline CODE_OF_CONDUCT.md]: I'd rephrase this as `The ZIO project is moderated by the Steering Committee team members:`.

File: CODE_OF_CONDUCT.md [hunk 2]
```diff
@@ -6,14 +6,31 @@ identity and expression, sexual orientation, disability, personal
...
+* John De Goes (@jdegoes)
+* Kai (@neko-kai)
+* Paul Shirshov (@pshirshov)
+* Salar Rahmanian (@softinio)
 
-For any questions, concerns, or moderation requests please contact a
-member of the project.
+The ZIO project requires that drastic moderation actions detailed in the code of 
+conduct - for example, removing a user from the Gitter channel - be agreed upon
+a group of over 2/3rds of the steering committee.
```
Comment (mijicd) [inline CODE_OF_CONDUCT.md]: It is kinda awkward to have 2/3 vote in a group of 5 :).

File: CODE_OF_CONDUCT.md [hunk 3]
```diff
@@ -6,14 +6,31 @@ identity and expression, sexual orientation, disability, personal
 appearance, body size, race, ethnicity, age, religion, nationality, or
 other such characteristics.
 
-Everyone is expected to follow the [Scala Code of Conduct] when
-discussing the project on the available communication channels. If you
-are being harassed, please contact us immediately so that we can
-support you.
+The ZIO project follows the [Scala Code of Conduct](https://www.scala-lang.org/conduct/), with
+adjustments with clarifications with regard to moderation that are detailed below.
```
Comment (mijicd) [inline CODE_OF_CONDUCT.md]: I'd replace `adjustment with ...` with something like `with an additional clause regarding moderation elaborated [here](section_link)`.

File: CODE_OF_CONDUCT.md (hunk 2 above)
Comment (hmemcpy) [reply CODE_OF_CONDUCT.md]: I propose adding Wiem to the steering committee.

Comment (jdegoes) [reply CODE_OF_CONDUCT.md]: /cc @wi101

**Comment by hmemcpy** [comment]
Not to detract from this important addition, I'm hearing concerns about the language of having a BDFL. Perhaps it would be better to elaborate on the extent and circumstances in which an override is required.

**Comment by softinio** [comment]
maybe we should consider a gitter channel and/or group email address to make it easier for people to contact the steering committee?

**Comment by jdegoes** [comment]
> Not to detract from this important addition, I'm hearing concerns about the language of having a BDFL.

Respectfully, I'm not playing the let's-get-consensus-with-everyone-on-Twitter-especially-the-ones-that-don't-like-JDG game. 😄

File: CODE_OF_CONDUCT.md [hunk 4]
```diff
@@ -6,14 +6,31 @@ identity and expression, sexual orientation, disability, personal
...
+conduct - for example, removing a user from the Gitter channel - be agreed upon
+a group of over 2/3rds of the steering committee.
 
-[Scala Code of Conduct]: https://www.scala-lang.org/conduct/
+For any questions, concerns, or moderation requests please contact any member of
+the project, or the people listed above.
+
+## BDFL
+
+In addition to the above, the ZIO project's BDFL (benevolent dictator for life) is
```
[the remark this answers is not in the document]
Comment (softinio) [reply CODE_OF_CONDUCT.md]: Word it without use of the bdfl title/wording? 

Other idea is for the purpose of coc we all take turns as chair of the steering committee with this responsibility when steering cannot reach a decision with a clear majority which should be very rare and hopefully never.

[the remark this answers is not in the document]
Comment (softinio) [reply CODE_OF_CONDUCT.md]: Word it without use of the bdfl title/wording? 

Other idea is for the purpose of coc we all take turns as chair of the steering committee with this responsibility when steering cannot reach a decision with a clear majority which should be very rare and hopefully never.

[the remark this answers is not in the document]
Comment (NeQuissimus) [reply CODE_OF_CONDUCT.md]: Just word it more neutral. John can be the tie breaker but the phrasing is just not good... "BDFL" sounds non-democratic and anti-communal. What we want is for John to support the steering committee, not be a dictator.

[the remark this answers is not in the document]
Comment (adamgfraser) [reply CODE_OF_CONDUCT.md]: Maybe Martin Odersky's role in the [Scala Improvement Process](https://docs.scala-lang.org/sips/sip-submission.html) and migration to Scala 3 is a good analogy, where he has a veto on decisions made by the committee? Hard to argue it is so terrible if it is used for governance of Scala itself.

[the remark this answers is not in the document]
Comment (jdegoes) [reply CODE_OF_CONDUCT.md]: "BDFL" is a safe-guard mechanism that would hopefully never be exercised, but it's important to state the expectations upfront.

If the steering committee wants to do one thing, and I want to do another, and we are unable to resolve our differences of opinion, what happens? We have to do something.

As projects become successful, they are subject to hostile takeovers. I saw that with Lars taking over the Scalaz project from Tony. I would not invest in this project like I have (untold hours and significant financial expense) if that were a remote possibility.

Having a BDFL will always rub people the wrong way, but it's a protection against things that matter to me very deeply—things that my founding and participation in this project have always been contingent on.

If I ever abuse BDFL for any reason, then it will be manifestly obvious, and I'd hope everyone I count as a friend would call me on that. I also hope that my long track record of keeping cool in any situation and never lashing out at anyone, no matter what they've done to me personally, will give people the confidence and trust they need to move forward.

[the remark this answers is not in the document]
Comment (ghostdogpr) [reply CODE_OF_CONDUCT.md]: Odersky himself is mentioned on the [Wikipedia page for BDFL](https://en.m.wikipedia.org/wiki/Benevolent_dictator_for_life) along with a lot of open source figures, so I don’t think it’s unreasonable John being BDFL for ZIO.

File: CODE_OF_CONDUCT.md (hunk 3 above)
Comment (regiskuckaertz) [inline CODE_OF_CONDUCT.md]:
```suggestion
adjustments and clarifications with regard to moderation that are detailed below.
```

File: CODE_OF_CONDUCT.md (hunk 2 above)
Comment (regiskuckaertz) [reply CODE_OF_CONDUCT.md]: crazy idea: 100%

File: CODE_OF_CONDUCT.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (softinio) [reply CODE_OF_CONDUCT.md]: Agree with @jdegoes lets with BDFL I say

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (softinio) [reply CODE_OF_CONDUCT.md]: Need to add @ghostdogpr to this list

Comment (softinio) [reply CODE_OF_CONDUCT.md]: and add @wi101

File: CODE_OF_CONDUCT.md [hunk 5]
```diff
@@ -34,3 +35,8 @@ In addition to the above, the ZIO project's BDFL (benevolent dictator for life)
 John De Goes (@jdegoes), owing to his original founding of the project and continued
 investments in it. While John adheres to the same code of conduct as everyone else,
 he is entitled to override moderation decisions made by the steering committee.
+
+We do not take the BDFL position lightly, especially with regards to moderation. John
+has consistently shown he is level-headed and able to handle conflict responsibly. Feel
+free to reach out to any member of the steering committee, including John himself,
+with any concern you might have.
```
Comment (hmemcpy) [inline CODE_OF_CONDUCT.md]: 👍

File: CODE_OF_CONDUCT.md (hunk 2 above)
Comment (iravid) [reply CODE_OF_CONDUCT.md]: I hope that 2/3rds of 7-8 members (rounded up this is 5 members) is a large enough majority to not be sufficiently different from 100%, while still allowing for progress in the face of disagreements.

File: CODE_OF_CONDUCT.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (iravid) [reply CODE_OF_CONDUCT.md]: I agree that the title "BDFL" has an anti-communal scent to it. But, I also trust John's ability to navigate conflict responsibly and to put ego aside. I also trust his ability to take criticism should we feel he abuses this position.

I had originally suggested this clause both for the above and because I am wary of hostile takeovers. Finally - I don't think we will need to exercise this clause, having 7-8 very able personas on board on the committee.

**Comment by iravid** [comment]
Leaving this open for another day. @zio/core, please re-review.

**Comment by softinio** [comment]
Shall we add a link to the code of conduct or have a duplicate of it on zio.dev ? We can tackle this as a separate PR if preferred so created #1611

File: CODE_OF_CONDUCT.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ioleo) [reply CODE_OF_CONDUCT.md]: @iravid After reading the wikipedia entry it seems like a good fit, but... it did sound like asking for drama until I learned that is an already estabilished term. Perhaps a quote from wikiepedia explaining the term or rephrasing to something more neutral, like `founding Father of ZIO` would serve us better in getting the "right message" across?

[the remark this answers is not in the document]
Comment (ioleo) [reply CODE_OF_CONDUCT.md]: @iravid 
Eg.
```
## Founding Father

In addition to the above the title Founding Father of ZIO project belongs to John De Goes (@jdegoes) indefinitely, owing to him originating the project and continued investments in it. While John adheres to the same code of conduct as everyone else, he is entitled to override moderation decisions made by the steering comittee.
```

File: CODE_OF_CONDUCT.md [hunk 6]
```diff
@@ -6,14 +6,37 @@ identity and expression, sexual orientation, disability, personal
...
 other such characteristics.
 
-Everyone is expected to follow the [Scala Code of Conduct] when
-discussing the project on the available communication channels. If you
-are being harassed, please contact us immediately so that we can
-support you.
+The ZIO project follows the [Scala Code of Conduct](https://www.scala-lang.org/conduct/), with
+an additional clause regarding moderation that is detailed below. All participants, contributors and 
+members are expected to follow the Scala Code of Conduct when discussing the project on the available 
+communication channels. If you are being harassed, please contact us immediately so that we can support you.
```
Comment (wi101) [inline CODE_OF_CONDUCT.md]: I would suggest if we generalize this:
```suggestion
communication channels. If the code of conduct is violated by any one, please contact one of the moderators immediately. 
```
or we can mention the discouraged behaviors to avoid that are mentioned in the CoC 
What do you think?

Comment (neko-kai) [reply CODE_OF_CONDUCT.md]: I think the previous one is less formulaic and IMHO a bit more useful, because while violation of a CoC is a *proxy* for harmful behavior, harm itself is primary. We shouldn't think about violating a CoC as an offense in itself, but rather about CoC being a prescription for the community on what constitutes and how to deal with harmful behavior.

Comment (wi101) [reply CODE_OF_CONDUCT.md]: Yes, you're right while it is mentioned that ZIO follows CoC it should be clear to respect all the statement there and the sentence like before makes sense too :).
